File: core/stumpff.py
import numpy as np

def stumpff_C(z):
    if z > 0:
        sqrt_z = np.sqrt(z)
        return (1 - np.cos(sqrt_z)) / z
    elif np.isclose(z, 0):
        return 0.5
    else:
        sqrt_neg_z = np.sqrt(-z)
        if sqrt_neg_z > 50:
            return 0.0
        return (1 - np.cosh(sqrt_neg_z)) / -z
    
def stumpff_S(z):
    if z > 0:
        sqrt_z = np.sqrt(z)
        return ((sqrt_z - np.sin(sqrt_z))) / (sqrt_z**3)
    elif np.isclose(z, 0):
        return 1/6
    else:
        sqrt_neg_z = np.sqrt(-z)
        if sqrt_neg_z > 50:
            return 0.0
        return (np.sinh(sqrt_neg_z) - sqrt_neg_z) / (sqrt_neg_z**3)

File: src/univ_prop/src_/stumpff.cpp
#include <cmath>

// Stumpff function C(z)
double stumpff_C(double z) {
    if (z > 0) {
        double sz = std::sqrt(z);
        return (1.0 - std::cos(sz)) / z;
    } else if (z < 0) {
        double sz = std::sqrt(-z);
        return (std::cosh(sz) - 1.0) / (-z);
    } else {
        return 0.5;
    }
}

// Stumpff function S(z)
double stumpff_S(double z) {
    if (z > 0) {
        double sz = std::sqrt(z);
        return (sz - std::sin(sz)) / (sz * sz * sz);
    } else if (z < 0) {
        double sz = std::sqrt(-z);
        return (std::sinh(sz) - sz) / (sz * sz * sz);
    } else {
        return 1.0 / 6.0;
    }
}